### PR TITLE
DEV2-3818 support insecure mode

### PR DESCRIPTION
**Pull request by amirbilu**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,470 +1,480 @@
 {
```
Comment (amirbilu) [inline package.json:1]: this wasn't prettiered

File: src/binary/runBinary.ts [hunk 2]
```diff
@@ -18,6 +18,8 @@ export default async function runBinary(
     : undefined;
   const args: string[] = [
     "--no-lsp=true",
+    "--tls_config",
+    `insecure=${tabnineExtensionProperties.ignoreCertificateErrors}`,
```
Comment (dimacodota) [inline src/binary/runBinary.ts:22]: this can't be merged before the binary is out

Comment (dimacodota) [reply src/binary/runBinary.ts:22]: what is the binary api for `tls_config` an array of flags?

File: src/enterprise/update/isHealthyServer.ts [hunk 3]
```diff
@@ -1,11 +1,12 @@
...
 
 export async function isHealthyServer(): Promise<boolean> {
   const url = serverUrl();
   if (!validateUrl(url)) {
     return false;
   }
-  const client = await createClient(url as string);
-  return healthy(client);
+
+  return (await getHttpStatusCode(new URL("/health", url).toString())) === 200;
```
Comment (dimacodota) [inline src/enterprise/update/isHealthyServer.ts:11]: extract to `ishealthy`

File: src/preRelease/installer.ts [hunk 4]
```diff
@@ -71,9 +69,7 @@ function isNewerAlphaVersionAvailable(
   const availableSemverCoerce = semver.coerce(availableVersion)?.version || "";
 
   const isNewerVersion =
-    !!currentVersion &&
-    semver.gt(availableVersion, currentVersion) &&
-    semver.neq(availableSemverCoerce, badVersion);
```
Comment (dimacodota) [inline src/preRelease/installer.ts:76]: why? not relevant?

File: src/proxyProvider.ts [hunk 5]
```diff
@@ -1,42 +1,41 @@
...
-    : undefined;
-  const port = Number.isNaN(parsedPort) ? undefined : parsedPort;
+  const proxyUrl = new URL(proxySettings);
 
-  const proxyOptions = {
-    host: proxyUrl.hostname,
-    port,
-    auth: proxyUrl.auth,
-    rejectUnauthorized,
+  const proxyOptions: HttpsProxyAgentOptions = {
```
Comment (dimacodota) [inline src/proxyProvider.ts:25]:
```suggestion
  const proxyOptions: HttpsProxyAgentOptions = {...proxyUrl, ca: options.ca, rejectUnauthorized: !options.ignoreCertificateErrors
```

File: src/utils/download.utils.ts [hunk 6]
```diff
@@ -44,20 +60,28 @@ function downloadResource<T>(
   ) => void
 ): Promise<T> {
   return new Promise<T>((resolve, reject) => {
-    const parsedUrl = url.parse(urlStr);
-    const { agent, rejectUnauthorized } = getHttpsProxyAgent();
-    const request: ClientRequest = getHttpClient(parsedUrl).request(
+    const parsedUrl = new URL(urlStr);
+    const request: ClientRequest = https.request(
```
Comment (dimacodota) [inline src/utils/download.utils.ts:64]: this won't work in  dev environment where you'r running a localhost  selfhosted service

Comment (dimacodota) [reply src/utils/download.utils.ts:64]: you need the  https://github.com/codota/tabnine-vscode/pull/1311/files#diff-d99f2ff5ebc521a8a2442af204589a4f49f486fef88b68e564d345b37ca0ad36L99

File: src/utils/download.utils.ts [hunk 7]
```diff
@@ -19,6 +34,7 @@ export function downloadFileToStr(urlStr: string): Promise<string> {
     });
   });
 }
+
 export function downloadFileToDestination(
   urlStr: string,
```
Comment (dimacodota) [inline src/utils/download.utils.ts:39]: change the api to support both, URl and string, and then you can drop the URL.toString()  in all the usages

File: src/utils/download.utils.ts [hunk 8]
```diff
@@ -44,20 +60,28 @@ function downloadResource<T>(
...
-    const request: ClientRequest = getHttpClient(parsedUrl).request(
+    const parsedUrl = new URL(urlStr);
+    const request: ClientRequest = https.request(
       {
-        host: parsedUrl.host,
-        path: parsedUrl.path,
-        port: getPortNumber(parsedUrl),
-        agent,
-        rejectUnauthorized,
+        protocol: parsedUrl.protocol,
```
Comment (dimacodota) [inline src/utils/download.utils.ts:66]: use "...parsedUrl"

File: src/utils/download.utils.ts [hunk 9]
```diff
@@ -87,15 +111,48 @@ function downloadResource<T>(
...
+      )
+      .on("error", (error) => {
+        reject(error);
+      });
+  });
 }
 
-function getHttpClient(parsedUrl: url.UrlWithStringQuery) {
-  return parsedUrl.protocol === "https:" ? https : http;
+export function readCaCertsSync(caCertsFileName: string): Buffer | undefined {
```
Comment (dimacodota) [inline src/utils/download.utils.ts:148]: why is it sync?

File: src/utils/download.utils.ts [hunk 10]
```diff
@@ -87,15 +111,48 @@ function downloadResource<T>(
...
-function getPortNumber(
-  parsedUrl: url.UrlWithStringQuery
-): string | number | undefined {
-  return (
-    (parsedUrl.port && Number(parsedUrl.port)) ||
-    (parsedUrl.protocol === "https:" ? 443 : 80)
-  );
+
+export function getHttpStatusCode(urlStr: string): Promise<number | undefined> {
+  return new Promise<number | undefined>((resolve, reject) => {
```
Comment (dimacodota) [inline src/utils/download.utils.ts:116]: why not just :
```
return downloadResource(urlStr, (response, resolve, reject) => {
    resolve(response.statusCode);
  });

File: src/binary/runBinary.ts (hunk 2 above)
Comment (amirbilu) [reply src/binary/runBinary.ts:22]: yes.. it currently supports --tls_config insecure=true ca_certs=path

File: src/enterprise/update/isHealthyServer.ts (hunk 3 above)
Comment (amirbilu) [reply src/enterprise/update/isHealthyServer.ts:11]: why? this is the name of the function (and no one else is using it)

File: src/preRelease/installer.ts (hunk 4 above)
Comment (amirbilu) [reply src/preRelease/installer.ts:76]: It's very old issue.. thought I can remove it already. I can re-add it if you think it's still needed.

File: src/proxyProvider.ts (hunk 5 above)
Comment (amirbilu) [reply src/proxyProvider.ts:25]: I tried it initially and it didn't work, not sure why .. you can try this in the browser console:
`{...new URL(window.location.href)}`

Comment (amirbilu) [reply src/proxyProvider.ts:25]: Apparently URL properties are not enumerable https://stackoverflow.com/questions/48232131/why-does-object-assign-not-copy-the-properties-of-a-url-object

File: src/utils/download.utils.ts (hunk 9 above)
Comment (amirbilu) [reply src/utils/download.utils.ts:148]: Will change

File: src/utils/download.utils.ts [hunk 11]
```diff
@@ -87,15 +111,48 @@ function downloadResource<T>(
...
-function getHttpClient(parsedUrl: url.UrlWithStringQuery) {
-  return parsedUrl.protocol === "https:" ? https : http;
+export function readCaCertsSync(caCertsFileName: string): Buffer | undefined {
+  try {
+    if (!caCertsFileName) {
+      return undefined;
+    }
+    return fs.readFileSync(caCertsFileName);
+  } catch (error) {
+    Logger.warn("Failed to read CA certs file", error);
```
Comment (yanir-codota) [inline src/utils/download.utils.ts:155]: For now we'll only see it in the logs, then?

File: src/enterprise/update/isHealthyServer.ts (hunk 3 above)
Comment (dimacodota) [reply src/enterprise/update/isHealthyServer.ts]: bc `ishealthy ` is readable ` (await getHttpStatusCode(new URL("/health", url).toString())) === 200` dose not

Comment (amirbilu) [reply src/enterprise/update/isHealthyServer.ts]: but the name of this function is `isHealthy`..

File: src/utils/download.utils.ts (hunk 11 above)
Comment (amirbilu) [reply src/utils/download.utils.ts:141]: yes.. this is not exposed for now

File: src/enterprise/update/isHealthyServer.ts (hunk 3 above)
Comment (dimacodota) [reply src/enterprise/update/isHealthyServer.ts]: oh, you meant the `isHealthyServer`, cool

**Pull request merged**